### PR TITLE
TailRec ADT with Loop/Done constructors to make intent more explicit

### PR DESCRIPTION
**Pull request by jacereda**

### CODE REVIEW

**Comment by paf31** [comment]
I like it, but I don't think the type constructor name `TailRec` is very instructive. That said, I can't think of many better options either. Maybe `Result`, or `Step`?

**Comment by jacereda** [comment]
I'm not sure what would look good in a type signature. Well, something like this would look good:

```
go :: Number -> Eff (trace :: Trace) (Number `StepsYielding` Unit)
```

but joking aside I guess Step is better than TailRec.

File: src/Control/Monad/Rec/Class.purs [hunk 1]
```diff
@@ -59,67 +63,71 @@ tailRecM3 f a b c = tailRecM (\o -> f o.a o.b o.c) { a: a, b: b, c: c }
...
-tailRec :: forall a b. (a -> Either a b) -> a -> b
+tailRec :: forall a b. (a -> Step a b) -> a -> b
 tailRec f a = go (f a)
   where
-  go (Left a) = go (f a)
-  go (Right b) = b
+  go (Loop a) = go (f a)
+  go (Done b) = b
+
+isLoop :: forall a b. Step a b -> Boolean
```
Comment (paf31) [inline src/Control/Monad/Rec/Class.purs]: Could you please move this into the `where` block?

File: src/Control/Monad/Rec/Class.purs [hunk 2]
```diff
@@ -20,6 +21,9 @@ import Control.Monad.Error.Trans
 import Control.Monad.State.Trans
 import Control.Monad.Writer.Trans
 
+
+data Step a b = Loop a | Done b
```
Comment (paf31) [inline src/Control/Monad/Rec/Class.purs:28]: Could you please add a markdown comment? Suggestion:

The result of a step in the computation: either `Loop`, specifying a new value for the accumulator, or indicate that the computation is `Done`, specifying a return value.

**Comment by safareli** [comment]
any updates on this?

**Comment by paf31** [comment]
I'd like to do this, but we'll need to wait until the next round of breaking changes since it affects lots of downstream libraries.

**Pull request closed without merging**